File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/EnterpriseCreatePortalFragment.java
package app.editors.manager.ui.fragments.login;

import android.os.Bundle;
import android.text.InputFilter;
import android.text.Spanned;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.AppCompatTextView;

import moxy.presenter.InjectPresenter;
import com.google.android.material.textfield.TextInputLayout;

import app.editors.manager.R;
import app.editors.manager.mvp.presenters.login.EnterpriseCreateValidatePresenter;
import app.editors.manager.mvp.views.login.EnterpriseCreateValidateView;
import app.editors.manager.ui.fragments.base.BaseAppFragment;
import app.editors.manager.ui.views.edits.BaseInputFilter;
import app.editors.manager.ui.views.edits.BaseWatcher;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import butterknife.Unbinder;
import lib.toolkit.base.managers.utils.StringUtils;
import lib.toolkit.base.managers.utils.UiUtils;
import lib.toolkit.base.ui.dialogs.common.CommonDialog;

public class EnterpriseCreatePortalFragment extends BaseAppFragment implements EnterpriseCreateValidateView {

    public static final String TAG = EnterpriseCreatePortalFragment.class.getSimpleName();

    @BindView(R.id.login_create_portal_address_edit)
    protected AppCompatEditText mLoginCreatePortalAddressEdit;
    @BindView(R.id.login_create_portal_address_hint_end)
    protected AppCompatTextView mLoginCreatePortalAddressHintEnd;
    @BindView(R.id.login_create_portal_address_edit_layout)
    protected TextInputLayout mLoginCreatePortalAddressLayout;
    @BindView(R.id.login_create_portal_email_edit)
    protected AppCompatEditText mLoginCreatePortalEmailEdit;
    @BindView(R.id.login_create_portal_email_layout)
    protected TextInputLayout mLoginCreatePortalEmailLayout;
    @BindView(R.id.login_create_portal_first_name_edit)
    protected AppCompatEditText mLoginCreatePortalFirstNameEdit;
    @BindView(R.id.login_create_portal_first_name_layout)
    protected TextInputLayout mLoginCreatePortalFirstNameLayout;
    @BindView(R.id.login_create_portal_last_name_edit)
    protected AppCompatEditText mLoginCreatePortalLastNameEdit;
    @BindView(R.id.login_create_portal_last_name_layout)
    protected TextInputLayout mLoginCreatePortalLastNameLayout;
    @BindView(R.id.login_signin_create_portal_button)
    protected AppCompatButton mLoginNextButton;

    @InjectPresenter
    EnterpriseCreateValidatePresenter mCreatePortalPresenter;

    private Unbinder mUnbinder;
    private FieldsWatcher mFieldsWatcher;
    private int mPaddingTop;
    private int mPaddingLeft;
    private int mPaddingRight;
    private int mPaddingBottom;

    public static EnterpriseCreatePortalFragment newInstance() {
        return new EnterpriseCreatePortalFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        final View view = inflater.inflate(R.layout.fragment_login_enterprise_create_portal, container, false);
        mUnbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init(savedInstanceState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }

    @OnClick(R.id.login_signin_create_portal_button)
    protected void onNextClick() {
        hideKeyboard(mLoginCreatePortalAddressEdit);
        final String address = mLoginCreatePortalAddressEdit.getText().toString();
        final String email = mLoginCreatePortalEmailEdit.getText().toString();
        final String first = mLoginCreatePortalFirstNameEdit.getText().toString();
        final String last = mLoginCreatePortalLastNameEdit.getText().toString();
        mCreatePortalPresenter.validatePortal(address, email, first, last);
    }

    @OnEditorAction(R.id.login_create_portal_last_name_edit)
    protected boolean actionKeyPress(final TextView v, final int actionId, final KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            onNextClick();
            return true;
        }

        return false;
    }

    @Override
    public void onCancelClick(CommonDialog.Dialogs dialogs, @Nullable final String tag) {
        super.onCancelClick(dialogs, tag);
        mCreatePortalPresenter.cancelRequest();
    }

    @Override
    public void onError(@Nullable String message) {
        hideDialog();
        showSnackBar(message);
    }

    @Override
    public void onValidatePortalSuccess() {
        hideDialog();
        showFragment(EnterpriseCreateSignInFragment.newInstance(),
                EnterpriseCreateSignInFragment.TAG, false);
    }

    @Override
    public void onPortalNameError(String message) {
        setEditHintVisibility(false);
        mLoginCreatePortalAddressLayout.setError(message);
    }

    @Override
    public void onEmailNameError(String message) {
        mLoginCreatePortalEmailLayout.setError(message);
    }

    @Override
    public void onFirstNameError(String message) {
        mLoginCreatePortalFirstNameLayout.setError(message);
    }

    @Override
    public void onLastNameError(String message) {
        mLoginCreatePortalLastNameLayout.setError(message);
    }

    @Override
    public void onRegionDomain(String domain) {
        final int textWidth = UiUtils.measureTextSizes(domain + "X", (int) mLoginCreatePortalAddressHintEnd.getTextSize()).x;
        mLoginCreatePortalAddressHintEnd.getLayoutParams().width = textWidth;
        mLoginCreatePortalAddressHintEnd.setText(domain);
        mPaddingRight = textWidth;
    }

    @Override
    public void onShowWaitingDialog(@StringRes int title) {
        showWaitingDialog(getString(title));
    }

    private void init(final Bundle savedInstanceState) {
        setActionBarTitle(getString(R.string.login_create_portal_title));
        showKeyboard(mLoginCreatePortalAddressEdit);
        mFieldsWatcher = new FieldsWatcher();
        mLoginNextButton.setEnabled(false);
        mLoginCreatePortalAddressEdit.requestFocus();
        mLoginCreatePortalAddressEdit.addTextChangedListener(mFieldsWatcher);
        mLoginCreatePortalAddressEdit.setFilters(new InputFilter[]{new FieldsFilter()});
        mLoginCreatePortalEmailEdit.addTextChangedListener(mFieldsWatcher);
        mLoginCreatePortalFirstNameEdit.addTextChangedListener(mFieldsWatcher);
        mLoginCreatePortalLastNameEdit.addTextChangedListener(mFieldsWatcher);
        mCreatePortalPresenter.getDomain();
        setPadding();
    }

    private void setPadding() {
        mPaddingTop = mLoginCreatePortalAddressEdit.getPaddingTop();
        mPaddingLeft = mLoginCreatePortalAddressEdit.getPaddingLeft();
        mPaddingRight = mLoginCreatePortalAddressEdit.getPaddingRight();
        mPaddingBottom = mLoginCreatePortalAddressEdit.getPaddingBottom();
    }

    private void setEditHintVisibility(final boolean isVisible) {
        if (isVisible) {
            mLoginCreatePortalAddressHintEnd.setVisibility(View.VISIBLE);
            mLoginCreatePortalAddressEdit.setPadding(mPaddingLeft, mPaddingTop, mPaddingRight, mPaddingBottom);
        } else {
            mLoginCreatePortalAddressHintEnd.setVisibility(View.GONE);
            mLoginCreatePortalAddressEdit.setPadding(mPaddingLeft, mPaddingTop, 0, mPaddingBottom);
        }
    }

    /*
     * Filtering inputs
     * */
    private class FieldsWatcher extends BaseWatcher {

        @Override
        public void onTextChanged(CharSequence text, int start, int before, int count) {
            mLoginCreatePortalEmailLayout.setErrorEnabled(false);
            mLoginCreatePortalFirstNameLayout.setErrorEnabled(false);
            mLoginCreatePortalLastNameLayout.setErrorEnabled(false);
            final String address = mLoginCreatePortalAddressEdit.getText().toString();
            final String email = mLoginCreatePortalEmailEdit.getText().toString();
            final String first = mLoginCreatePortalFirstNameEdit.getText().toString();
            final String last = mLoginCreatePortalLastNameEdit.getText().toString();

            if (StringUtils.isAlphaNumeric(address)) {
                mLoginCreatePortalAddressLayout.setErrorEnabled(false);
            }

            if (!"".equals(first) & StringUtils.isCreateUserName(first)) {
                mLoginCreatePortalFirstNameLayout.setError(getString(R.string.errors_first_name));
                mLoginNextButton.setEnabled(false);
            } else if (!"".equals(last) & StringUtils.isCreateUserName(last)) {
                mLoginCreatePortalLastNameLayout.setError(getString(R.string.errors_last_name));
                mLoginNextButton.setEnabled(false);
            } else if (!"".equals(address) && !"".equals(email) && !"".equals(first) && !"".equals(last) && StringUtils.isAlphaNumeric(address)) {
                mLoginNextButton.setEnabled(true);
            } else {
                mLoginNextButton.setEnabled(false);
            }

        }
    }

    private class FieldsFilter extends BaseInputFilter {

        @Override
        public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
            super.filter(source, start, end, dest, dstart, dend);
            if (mCreatePortalPresenter.checkPhrase(mResultString)) {
                return null;
            }

            if (!StringUtils.isAlphaNumeric(mResultString)) {
                mLoginCreatePortalAddressLayout.setError(getString(R.string.login_api_portal_name_content));
                return source;
            } else {
                mLoginCreatePortalAddressLayout.setErrorEnabled(false);
                setEditHintVisibility(true);
                return null;
            }
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/EnterpriseCreatePortalFragment.kt
package app.editors.manager.ui.fragments.login

import android.content.Context
import android.os.Bundle
import android.text.InputFilter
import android.text.Spanned
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.fragment.app.viewModels
import app.editors.manager.R
import app.editors.manager.app.appComponent
import app.editors.manager.databinding.FragmentLoginEnterpriseCreatePortalBinding
import app.editors.manager.ui.fragments.base.BaseAppFragment
import app.editors.manager.ui.views.edits.BaseInputFilter
import app.editors.manager.ui.views.edits.BaseWatcher
import app.editors.manager.viewModels.login.CreatePortalState
import app.editors.manager.viewModels.login.EnterpriseCreateValidateViewModel
import lib.toolkit.base.managers.utils.StringUtils.isAlphaNumeric
import lib.toolkit.base.managers.utils.StringUtils.isCreateUserName
import lib.toolkit.base.managers.utils.UiUtils.measureTextSizes
import lib.toolkit.base.ui.dialogs.common.CommonDialog.Dialogs

class EnterpriseCreatePortalFragment : BaseAppFragment() {

    companion object {
        val TAG: String = EnterpriseCreatePortalFragment::class.java.simpleName
        fun newInstance(): EnterpriseCreatePortalFragment {
            return EnterpriseCreatePortalFragment()
        }
    }

    private val viewModel by viewModels<EnterpriseCreateValidateViewModel>()

    private var viewBinding: FragmentLoginEnterpriseCreatePortalBinding? = null

    private var fieldsWatcher: FieldsWatcher? = null
    private var paddingTop = 0
    private var paddingLeft = 0
    private var paddingRight = 0
    private var paddingBottom = 0

    override fun onAttach(context: Context) {
        super.onAttach(context)
        requireContext().appComponent.inject(viewModel)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        viewBinding = FragmentLoginEnterpriseCreatePortalBinding.inflate(inflater)
        return viewBinding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }


    private fun onNextClick() {
        hideKeyboard(viewBinding?.loginCreatePortalAddressEdit)
        val address = viewBinding?.loginCreatePortalAddressEdit?.text.toString()
        val email = viewBinding?.loginCreatePortalEmailEdit?.text.toString()
        val first = viewBinding?.loginCreatePortalFirstNameEdit?.text.toString()
        val last = viewBinding?.loginCreatePortalLastNameEdit?.text.toString()
        viewModel.validatePortal(address, email, first, last)
    }


    private fun actionKeyPress(v: TextView?, actionId: Int, event: KeyEvent?): Boolean {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            onNextClick()
            return true
        }
        return false
    }

    override fun onCancelClick(dialogs: Dialogs?, tag: String?) {
        super.onCancelClick(dialogs, tag)
        viewModel.cancelRequest()
    }

    fun onError(message: String?) {
        hideDialog()
        message?.let { showSnackBar(message) }
    }

    private fun onValidatePortalSuccess(email: String?, first: String?, last: String?) {
        hideDialog()
        showFragment(
            EnterpriseCreateSignInFragment.newInstance(email, first, last),
            EnterpriseCreateSignInFragment.TAG, false
        )
    }

    private fun onPortalNameError(message: String) {
        setEditHintVisibility(false)
        viewBinding?.loginCreatePortalAddressEditLayout?.error = message
    }

    private fun onEmailNameError(message: String) {
        viewBinding?.loginCreatePortalEmailLayout?.error = message
    }

    private fun onFirstNameError(message: String) {
        viewBinding?.loginCreatePortalFirstNameLayout?.error = message
    }

    private fun onLastNameError(message: String) {
        viewBinding?.loginCreatePortalLastNameLayout?.error = message
    }

    private fun onRegionDomain(domain: String) {
        val textWidth = measureTextSizes(
            domain + "X", viewBinding?.loginCreatePortalAddressHintEnd?.textSize?.toInt() ?: -1
        ).x
        viewBinding?.loginCreatePortalAddressHintEnd?.apply {
            layoutParams.width = textWidth
            text = domain
        }
        paddingRight = textWidth
    }

    private fun onShowWaitingDialog(@StringRes title: Int) {
        showWaitingDialog(getString(title))
    }

    private fun init() {
        fieldsWatcher = FieldsWatcher()
        initListeners()
        setActionBarTitle(getString(R.string.login_create_portal_title))
        showKeyboard(viewBinding?.loginCreatePortalAddressEdit)
        viewBinding?.loginSigninCreatePortalButton?.isEnabled = false
        viewBinding?.loginCreatePortalAddressEdit?.apply {
            requestFocus()
            filters = arrayOf<InputFilter>(FieldsFilter())
        }
        setPadding()
        initViewModel()
    }

    private fun initViewModel() {
        viewModel.errorLiveData.observe(viewLifecycleOwner) { error ->
            onError(error.message)
        }
        viewModel.regionLiveData.observe(viewLifecycleOwner) { domain ->
            onRegionDomain(domain ?: "")
        }
        viewModel.stateLiveData.observe(viewLifecycleOwner) { state ->
            when (state) {
                is CreatePortalState.Success -> {
                    onValidatePortalSuccess(
                        state.portalModel.email,
                        state.portalModel.firstName,
                        state.portalModel.lastName
                    )
                }
                is CreatePortalState.Error -> {
                    checkError(state.res)
                }
                is CreatePortalState.Progress -> {
                    onShowWaitingDialog(R.string.dialogs_wait_title)
                }
            }
        }
        viewModel.getDomain()
    }

    private fun checkError(errorMessage: Int?) {
        errorMessage?.let { message ->
            when (message) {
                R.string.login_api_portal_name_length -> {
                    onPortalNameError(getString(message))
                }
                R.string.errors_email_syntax_error -> {
                    onEmailNameError(getString(message))
                }
                R.string.errors_first_name -> {
                    onFirstNameError(getString(message))
                }
                R.string.errors_last_name -> {
                    onLastNameError(getString(message))
                }
                else -> onError(getString(message))
            }
        } ?: onError(null)
    }

    private fun initListeners() {
        viewBinding?.loginSigninCreatePortalButton?.setOnClickListener {
            onNextClick()
        }

        viewBinding?.loginCreatePortalLastNameEdit?.setOnEditorActionListener { v, actionId, event ->
            actionKeyPress(v, actionId, event)
        }
        viewBinding?.loginCreatePortalAddressEdit?.addTextChangedListener(fieldsWatcher)
        viewBinding?.loginCreatePortalEmailEdit?.addTextChangedListener(fieldsWatcher)
        viewBinding?.loginCreatePortalFirstNameEdit?.addTextChangedListener(fieldsWatcher)
        viewBinding?.loginCreatePortalLastNameEdit?.addTextChangedListener(fieldsWatcher)
    }

    private fun setPadding() {
        paddingTop = viewBinding?.loginCreatePortalAddressEdit?.paddingTop ?: 0
        paddingLeft = viewBinding?.loginCreatePortalAddressEdit?.paddingLeft ?: 0
        paddingRight = viewBinding?.loginCreatePortalAddressEdit?.paddingRight ?: 0
        paddingBottom = viewBinding?.loginCreatePortalAddressEdit?.paddingBottom ?: 0
    }

    private fun setEditHintVisibility(isVisible: Boolean) {
        if (isVisible) {
            viewBinding?.loginCreatePortalAddressHintEnd?.apply {
                visibility = View.VISIBLE
                setPadding(
                    paddingLeft,
                    paddingTop,
                    paddingRight,
                    paddingBottom
                )
            }
        } else {
            viewBinding?.loginCreatePortalAddressHintEnd?.apply {
                visibility = View.GONE
                setPadding(paddingLeft, paddingTop, 0, paddingBottom)
            }
        }
    }

    /*
     * Filtering inputs
     * */
    private inner class FieldsWatcher : BaseWatcher() {
        override fun onTextChanged(text: CharSequence, start: Int, before: Int, count: Int) {
            viewBinding?.loginCreatePortalEmailLayout?.isErrorEnabled = false
            viewBinding?.loginCreatePortalFirstNameLayout?.isErrorEnabled = false
            viewBinding?.loginCreatePortalLastNameLayout?.isErrorEnabled = false
            val address = viewBinding?.loginCreatePortalAddressEdit?.text.toString()
            val email = viewBinding?.loginCreatePortalEmailEdit?.text.toString()
            val first = viewBinding?.loginCreatePortalFirstNameEdit?.text.toString()
            val last = viewBinding?.loginCreatePortalLastNameEdit?.text.toString()
            if (isAlphaNumeric(address)) {
                viewBinding?.loginCreatePortalAddressEditLayout?.isErrorEnabled = false
            }
            when {
                ("" != first) and isCreateUserName(first) -> {
                    viewBinding?.loginCreatePortalFirstNameLayout?.error = getString(R.string.errors_first_name)
                    viewBinding?.loginSigninCreatePortalButton?.isEnabled = false
                }
                ("" != last) and isCreateUserName(last) -> {
                    viewBinding?.loginCreatePortalLastNameLayout?.error = getString(R.string.errors_last_name)
                    viewBinding?.loginSigninCreatePortalButton?.isEnabled = false
                }
                else -> viewBinding?.loginSigninCreatePortalButton?.isEnabled =
                    "" != address && "" != email && "" != first && "" != last && isAlphaNumeric(
                        address
                    )
            }
        }
    }

    private inner class FieldsFilter : BaseInputFilter() {
        override fun filter(
            source: CharSequence,
            start: Int,
            end: Int,
            dest: Spanned,
            dstart: Int,
            dend: Int
        ): CharSequence? {
            super.filter(source, start, end, dest, dstart, dend)
            if (viewModel.checkPhrase(mResultString)) {
                return null
            }
            return if (!isAlphaNumeric(mResultString)) {
                viewBinding?.loginCreatePortalAddressEditLayout?.error =
                    getString(R.string.login_api_portal_name_content)
                source
            } else {
                viewBinding?.loginCreatePortalAddressEditLayout?.isErrorEnabled = false
                setEditHintVisibility(true)
                null
            }
        }
    }
}